### PR TITLE
Issue 607

### LINKED ISSUE #607: Define permission level for ActiveWindows
### What steps will reproduce the problem?
In a model, I define an acvitewindow like this
public function ngRestActiveWindows()
    {
        return [
            ['class' => 'luya\admin\aws\DetailViewActiveWindow', 
                                   'attributes' => [
                                                                   ...,  //Attributes I want to show up        
                                                            ]
           ]
        ];
    }

In the permissions rights of a user group, I just set the view permission right.
### What is the expected result?
Button for the DetailViewActiveWindow should be shown 

### What do you get instead? (A Screenshot can help us a lot!)
No buttons shown, the detail button only shown if at least Update permissions rights is set for the user group.

Watch this for more clear: https://youtu.be/MivmTFgCV6M

### PR DESCRIPTION
**Pull request by rochdi80tn**
### What are you changing/introducing
Define permission level for ActiveWindows and ActiveButtons
  ```
public function ngRestActiveWindows()
 {
    return [
        ['class' => ChangePasswordActiveWindow::class, 'label' => false, 'permissionLevel' => Auth:CAN_VIEW],
    ];
}
```

Same logic could be applied to `ngRestActiveButtons()` too

```
public function ngRestActiveButtons()
{
    return [
        ['class' => TimestampActiveButton::class, 'attribute' => 'foo', 'permissionLevel' => Auth:CAN_UPDATE]
    ];
}
```

### What is the reason for changing/introducing
Active buttons and windows are only displayed when the user had the UPDATE permission.
With this change, we can defined the permisison level we want for buttons to be displayed.
If none is defined, buttons will be displayed always.

### QA

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes
| Tests pass?   | yes
| Fixed issues  | closes #607

### CODE REVIEW

File: src/ngrest/Config.php [hunk 1]
```diff
@@ -1,660 +1,661 @@
...
-                    'type' => [
-                        'class' => 'luya\admin\ngrest\plugins\Number',
-                        'args' => [],
-                    ],
-                ]);
-            }
-        }
-    }
-}
+<?php
```
Comment (nadar) [inline src/ngrest/Config.php:1]: @rochdi80tn can you explain me why every line has changes, but you have NOT changed anything? This looks strange.

Comment (nadar) [reply src/ngrest/Config.php:1]: Have you set tabs as tabs instead of 4 spaces maybe?

File: src/ngrest/NgRestPermissionLevelInterface.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * Ngrest Button Condition Interface
+ *
+ * @author Rochdi Bazine <rochdi80tn@gmail.com>
+ * @since 4.0.0
+ */
+interface NgRestPermissionLevelInterface
+{
+    /**
+     * Setter method for the ActiveWindow button displaying permission level
+     * @param int $permissionLevel
```
Comment (nadar) [inline src/ngrest/NgRestPermissionLevelInterface.php:15]: Maybe extend this param doc by saying, use `Auth::CAN_VIEW, ...`

File: src/ngrest/NgRestPermissionLevelInterface.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace luya\admin\ngrest;
+
+/**
+ * Ngrest Button Condition Interface
```
Comment (nadar) [inline src/ngrest/NgRestPermissionLevelInterface.php:6]: Ngrest Button Permission Condition Interface

File: src/ngrest/base/ActiveButton.php [hunk 4]
```diff
@@ -154,7 +155,29 @@ public function getCondition()
     {
         return empty($this->_condition) ? '' : $this->_condition;
     }
-
+    
+    private $_permissionLevel = null;
+    
+    /**
```
Comment (nadar) [inline src/ngrest/base/ActiveButton.php:161]: When its a method from the interface you can use write `{@inheritdoc}` then IDE's will use the inherited doc and also the luya.io api renderer.

Comment (nadar) [reply src/ngrest/base/ActiveButton.php:161]:
```
/**
 * {@inheritDoc}
 */
```

Is enough then.

File: src/ngrest/base/ActiveButton.php [hunk 5]
```diff
@@ -154,7 +155,29 @@ public function getCondition()
...
+    {
+        $this->_permissionLevel = $permissionLevel;
+    }
+    
+    /**
+     * Get the button displaying permission level or empty of not set
+     * @inheritdoc
+     * @since 4.0.0
+     */
+    public function getPermissionLevel()
```
Comment (nadar) [inline src/ngrest/base/ActiveButton.php:176]:
```
/**
 * {@inheritDoc}
 */
```

File: src/ngrest/base/ActiveButton.php [hunk 6]
```diff
@@ -154,7 +155,29 @@ public function getCondition()
...
+    {
+        return $this->_permissionLevel === null ? null : $this->_permissionLevel;
```
Comment (nadar) [inline src/ngrest/base/ActiveButton.php:178]: The oppsite would be less complicate:

```
return $this->_permissionLevel ? $this->_permissionLevel : null;
```

Or is there a need to explicit check for `null`?

File: src/ngrest/base/ActiveButton.php [hunk 7]
```diff
@@ -154,7 +155,29 @@ public function getCondition()
     {
         return empty($this->_condition) ? '' : $this->_condition;
     }
-
+    
+    private $_permissionLevel = null;
```
Comment (nadar) [inline src/ngrest/base/ActiveButton.php:159]: private $_permissionLevel;

is equals to

private $_permissionLevel = null;

So no need for default value definition.

File: src/ngrest/base/ActiveWindow.php [hunk 8]
```diff
@@ -375,6 +376,28 @@ public function getCondition()
         return empty($this->_condition) ? '' : $this->_condition;
     }
     
+    private $_permissionLevel = null;
```
Comment (nadar) [inline src/ngrest/base/ActiveWindow.php:379]: private $_permissionLevel;

is equals to

private $_permissionLevel = null;

So no need for default value definition.

File: src/ngrest/base/ActiveWindow.php [hunk 9]
```diff
@@ -375,6 +376,28 @@ public function getCondition()
         return empty($this->_condition) ? '' : $this->_condition;
     }
     
+    private $_permissionLevel = null;
+    
+    /**
+     * Setter method for the ActiveWindow button displaying permission level
+     * @param int $permissionLevel
+     * @since 4.0.0
```
Comment (nadar) [inline src/ngrest/base/ActiveWindow.php:384]:
```
/**
 * {@inheritDoc}
 */
```

File: src/ngrest/base/ActiveWindow.php [hunk 10]
```diff
@@ -375,6 +376,28 @@ public function getCondition()
...
+     */
+    public function setPermissionLevel($permissionLevel)
+    {
+        $this->_permissionLevel = $permissionLevel;
+    }
+    
+    /**
+     * Get the button displaying permission level or empty of not set
+     * @inheritdoc
+     * @since 4.0.0
```
Comment (nadar) [inline src/ngrest/base/ActiveWindow.php:394]:
```
/**
 * {@inheritDoc}
 */
```

File: src/ngrest/base/ActiveWindow.php [hunk 11]
```diff
@@ -375,6 +376,28 @@ public function getCondition()
...
+     */
+    public function getPermissionLevel()
+    {
+        return $this->_permissionLevel === null  ? null : $this->_permissionLevel;
```
Comment (nadar) [inline src/ngrest/base/ActiveWindow.php:398]: Same here for null check. and there are two whitespaces behind === null

**Review by nadar — CHANGES_REQUESTED** [review]
Some small changes, overall looks good. thanks!

File: src/ngrest/Config.php (hunk 1 above)
Comment (rochdi80tn) [reply src/ngrest/Config.php:1]: Not sure.. I am using a linux shell WSL mode for git commits.. 
Hadn't messed up with tabs and spaces..
I'll double check this odd behavior

File: src/ngrest/NgRestPermissionLevelInterface.php (hunk 3 above)
Comment (rochdi80tn) [reply src/ngrest/NgRestPermissionLevelInterface.php:6]: Ok

File: src/ngrest/base/ActiveButton.php (hunk 4 above)
Comment (rochdi80tn) [reply src/ngrest/base/ActiveButton.php:161]: OK

File: src/ngrest/base/ActiveButton.php (hunk 5 above)
Comment (rochdi80tn) [reply src/ngrest/base/ActiveButton.php:176]: OK

File: src/ngrest/base/ActiveButton.php (hunk 6 above)
Comment (rochdi80tn) [reply src/ngrest/base/ActiveButton.php:178]: Permission level is an integer including 0 value (Auth::CAN_VIEW) so actually I am checking explicitly if it is null to know that it has not been initialized..

File: src/ngrest/base/ActiveWindow.php (hunk 8 above)
Comment (rochdi80tn) [reply src/ngrest/base/ActiveWindow.php:379]: Ok, wanted to be sure to default the value to null.. wasn't sure about the default value if no thing is specified. So will update accordingly.

Comment (rochdi80tn) [reply src/ngrest/base/ActiveWindow.php:379]: @nadar this actually makes me wondering if it wouldn't be good to set the default permission level to Auth::CAN_UPDATE this way we will not break backward compatibility.. What do you suggest ?

Comment (nadar) [reply src/ngrest/base/ActiveWindow.php:379]: Thanks for pointing into that direction, yes indeed we should set can_update in order to prevent BC! Thanks

**Comment by rochdi80tn** [comment]
@nadar, can you help fixing this testing error https://github.com/luyadev/luya-module-admin/pull/630/checks?check_run_id=2220290283#step:7:78.. Looks not related to my commits. I might be wrong though

**Comment by rochdi80tn** [comment]
> @nadar, can you help fixing this testing error https://github.com/luyadev/luya-module-admin/pull/630/checks?check_run_id=2220290283#step:7:78.. Looks not related to my commits. I might be wrong though

@nadar, was my mistake I miss replaced a null value by CAN_UPDATE.. should be fixed with latest commit as of this

**Comment by rochdi80tn** [comment]
This can be added  docs/guide/ngrest-activewindow.md doc:


#### Button permission level

In addition to button condition a permission level can be set. This will ensure displaying the buttons only when proper permission level is met.
This can be defined with the `permissionLevel` atrribute in `ngRestActiveWindows()` and `ngRestActiveButtons()` model definitions. Example:

```php
public function ngRestActiveWindows()
{
    return [
        ['class' => \luya\admin\aws\TestActiveWindow::className(), 'label' => 'My Window ', 'permissionLevel' => CAN_VIEW],
    ];
}
```
Above example will aadd the TestActiveWindow button if the user has the VIEW permission of the model.

If `permissionLevel` attribute is not set, the default behavior is applied which is `CAN_UPDATE` permission is required.

To disable permission checks, you have to set the `permissionLevel` to am empty string : 
```php
...
        ['class' => \luya\admin\aws\TestActiveWindow::className(), 'label' => 'My Window ', 'permissionLevel' => ''],
...
```

**Review by nadar — CHANGES_REQUESTED** [review]
Only the changelog entry is missing. Thank you! Its perfect

**Comment by nadar** [comment]
Guide has been added https://github.com/luyadev/luya/commit/d0571b5ac0129c511ade82c486e55b382ae1189c

**Review by nadar — APPROVED** [review]

**Pull request merged**